tss-esapi/src/abstraction/no_tpm/quote.rs

```diff
@@ -0,0 +1,379 @@
+// Copyright 2021 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use crate::error::Error;
+use crate::error::Result;
+use crate::WrapperErrorKind;
+use crate::{
+    interface_types::algorithm::HashingAlgorithm,
+    structures::{Attest, AttestInfo, DigestList, PcrSelectionList, Public, QuoteInfo, Signature},
+    traits::Marshall,
+};
+use digest::{Digest, DynDigest};
+
+#[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]
+use crate::{abstraction::public::AssociatedTpmCurve, structures::EccSignature};
+#[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]
+use ecdsa::{
+    hazmat::{DigestPrimitive, VerifyPrimitive},
+    PrimeCurve, SignatureSize, VerifyingKey,
+};
+#[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]
+use elliptic_curve::{
+    generic_array::ArrayLength,
+    point::AffinePoint,
+    sec1::{FromEncodedPoint, ModulusSize, ToEncodedPoint},
+    CurveArithmetic, FieldBytesSize,
+};
+#[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]
+use signature::hazmat::PrehashVerifier;
+
+#[cfg(feature = "rsa")]
+use rsa::{pkcs1v15, pss, RsaPublicKey};
+#[cfg(feature = "rsa")]
+use signature::Verifier;
+
+#[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]
+fn verify_ecdsa<C>(
+    public: &Public,
+    message: &[u8],
+    signature: &EccSignature,
+    hashing_algorithm: HashingAlgorithm,
+) -> Result<bool>
+where
+    C: PrimeCurve + CurveArithmetic + DigestPrimitive + AssociatedTpmCurve,
+    AffinePoint<C>: VerifyPrimitive<C> + FromEncodedPoint<C> + ToEncodedPoint<C>,
+    SignatureSize<C>: ArrayLength<u8>,
+    FieldBytesSize<C>: ModulusSize,
+{
+    let Ok(signature) = ecdsa::Signature::<C>::try_from(signature.clone()) else {
+        return Ok(false);
+    };
+    let Ok(public) = elliptic_curve::PublicKey::<C>::try_from(public) else {
+        return Ok(false);
+    };
+
+    let verifying_key = VerifyingKey::from(public);
+
+    match hashing_algorithm {
+        #[cfg(feature = "sha1")]
+        HashingAlgorithm::Sha1 => {
+            let hash = sha1::Sha1::digest(message);
+            Ok(verifying_key.verify_prehash(&hash, &signature).is_ok())
```

Comment:
The use of `PrehashVerifier` should be swapped with `DigestVerifier` just so we stay off the hazmat.

https://github.com/RustCrypto/elliptic-curves/pull/1146